lucene/core/src/java/org/apache/lucene/util/crypto/EncryptedFileChannel.java

```diff
@@ -0,0 +1,312 @@
+      return options;
+    }
+    optList.add(StandardOpenOption.WRITE);
+    return optList.toArray(new OpenOption[optList.size()]);
+  }
+
+  private final IvParameterSpec getIv(OpenOption... options) throws IOException {
+    if (this.channel.size() >= IV_LENGTH) {
+      // non-empty file, assume IV is present at the beginning
+      ByteBuffer tmp = ByteBuffer.allocate(IV_LENGTH);
+      this.channel.read(tmp);
+      
+      this.ivLength = IV_LENGTH;
+      return new IvParameterSpec(tmp.array());
+    } else if (Arrays.asList(options).contains(StandardOpenOption.WRITE)) {
+      // empty file, generate IV
+      IvParameterSpec iv;
+      try {
+        iv = Crypto.generateAesIV();
+      } catch (NoSuchAlgorithmException e) {
+        throw new IOException(e);
+      }
+      // Don't write IV yet (so in case this channel is newer written it will remain completely empty)
+      this.ivLength = 0;
+      return iv;
+    } else {
+      // empty file but not writable (can happen in tests etc.)
+      return null;
+    }
+  }
+  
+  private void ensureIv() throws IOException {
+    if (cipher == null || ivWritten) {
+      return;
+    }
+    assert this.channel.position() == 0 : "Bytes written before IV! " + this.channel.position();
+    this.ivWritten = true;
+    ByteBuffer tmp = ByteBuffer.allocate(IV_LENGTH);
+    tmp.put(cipher.getIV());
+    tmp.flip();
+    this.channel.write(tmp);
+    this.ivLength = IV_LENGTH;
+  }
+
+  @Override
+  public int read(ByteBuffer dst) throws IOException {
+    ByteBuffer tmp = ByteBuffer.allocate(dst.limit() - dst.position());
+    long position = this.position();
+    // Note that cipher may be null but only when the channel is empty so we expect
+    // to never come here.
+    assert cipher == null || position >= 0 : "Cannot read EncryptedFileChannel before IV: " + this.channel.position();
+    int read = this.channel.read(tmp);
+    if (read <= 0) {
+      return read;
+    }
+    tmp.flip();
+    dst.put(this.cipher.decrypt(tmp, position), 0, read);
+
+    return read;
+  }
+
+  @Override
+  public int read(ByteBuffer dst, long position) throws IOException {
+    ByteBuffer tmp = ByteBuffer.allocate(dst.limit() - dst.position());
+    int read = this.channel.read(tmp, position + ivLength);
+    if (read <= 0) {
+      return read;
+    }
+    tmp.flip();
+    dst.put(this.cipher.decrypt(tmp, position), 0, read);
+
+    return read;
+  }
+
+  @Override
+  public long read(ByteBuffer[] dsts, int offset, int length) throws IOException {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public int write(ByteBuffer src) throws IOException {
+    if (this.cipher == null) {
+      // in case of empty readonly channel - let it crash on write
+      return this.channel.write(src);
+    }
+    ensureIv();
+    long position = this.position();
+    assert position >= 0 : "Cannot write EncryptedFileChannel before IV: " + this.channel.position();
```

Comment:
Shouldn't the assertion be `position >= IV_LENGTH`? So that you can't write unless there is already an IV written at the beginning of the file.

Reply:
No, because `position` comes from `this.position()` which already substracts the `IV_LENGTH`.
But you have a point - this assertion does nothing, I'll rewrite it.

Reply:
No, it does something - `this.position()` would return a negative number if we already have written the IV but the the position was changed (by a third party) in `this.channel()` to a one whithin the IV